icu4c/source/common/unicode/char16ptr.h

```diff
@@ -334,6 +352,21 @@ inline std::u16string_view toU16StringView(std::wstring_view sv) {
     return { ConstChar16Ptr(sv.data()), sv.length() };
 }
 #endif
+
+template <typename T,
+          typename = typename std::enable_if_t<!std::is_pointer_v<std::remove_reference_t<T>>>>
+inline std::u16string_view toU16StringViewNullable(const T& text) {
+    return toU16StringView(text);
+}
+
+template <typename T,
+          typename = typename std::enable_if_t<std::is_pointer_v<std::remove_reference_t<T>>>,
+          typename = void>
```

Comment:
just curious: what does this last `typename = void` do?

Reply:
Oh, this is a classic ugly trick stemming from the fact that enable_if is an ugly trick.
The previous one is a declaration of
`template<typename T, typename (unnamed)> std::u16string_view toU16StringViewNullable(const T&)`.
You cannot declare the same thing a second time with a different default template parameter (even though the intended usage here is that you never pass the unnamed template parameter, and that the defaults cause only one of the two to exist). So you declare something different,
`template<typename T, typename (unnamed), typename (also unnamed)> std::u16string_view toU16StringViewNullable(const T&)`.

This will all go away with C++20 and requires clauses, and then the requires clause will be part of what you declare and a small amount of sanity will be restored.

Reply:
See https://godbolt.org/z/eGno9xrsP.

Reply:
And for completeness, a glimpse into the bright future of the year 2020: https://godbolt.org/z/e4G48W36n.